program-analysis/echidna/Exercise-X.md

```diff
@@ -0,0 +1,442 @@
+            // actions
+            naughtCoin.transfer(bob, amount);
+        }
+        // post-conditions
+        assert(
+            naughtCoin.balanceOf(player) == playerBalanceBefore &&
+                naughtCoin.balanceOf(bob) == bobBalanceBefore
+        );
+    }
+   ```
+
+Run Echidna and check the corpus. Is this a good property?
+
+</details>
+
+<details>
+  <summary>Improving the 3rd property</summary>
+
+   If you look at the corpus:
+
+   ```
+ 34 | r   |     function transfer_should_fail_before_timelock_period() public {
+ 35 |     |         // pre-conditions
+ 36 | r   |         uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
+ 37 | r   |         uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
+ 38 | r   |         uint256 currentTime = block.timestamp;
+ 39 | r   |         if (currentTime < naughtCoin.timeLock()) {
+ 40 |     |             // actions
+ 41 | r   |             naughtCoin.transfer(bob, 100);
+ 42 |     |         }
+ 43 |     |         // post-conditions
+ 44 |     |         assert(
+ 45 |     |             naughtCoin.balanceOf(player) == playerBalanceBefore &&
+ 46 |     |                 naughtCoin.balanceOf(bob) == bobBalanceBefore
+ 47 |     |         );
+ 48 |     |     }
+ 49 |     | }
+   ```
+
+   You can see that the transfer reverted as expected. But... Our post-conditions weren't checked. If you have read the [ERC20 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md), you will know that the `transfer` function returns a status `boolean`. In the case of success, we can emit a special `AssertionFailed` event that won't stop the execution of the function. This way, our post-conditions will be checked.
+
+- If `transfer` failed, that's okay. Keep going.
+- If `transfer` succeeded, we don't want that, emit the `AssertionFailed` and check post-conditions as usual.
+ 
+ Add the `event AssertionFailed(uint256 amount);` at the top of your contract.
+ 
+   ```
+   function transfer_should_fail_before_timelock_period(uint256 amount)
+        public
+    {
+        // pre-conditions
+        uint256 playerBalanceBefore = naughtCoin.balanceOf(player);
+        uint256 bobBalanceBefore = naughtCoin.balanceOf(bob);
+        uint256 currentTime = block.timestamp;
+        if (currentTime < naughtCoin.timeLock()) {
+            // actions
+            bool success1 = naughtCoin.transfer(bob, amount);
+            if (success1) {
+                emit AssertionFailed(amount);
+            }
+        }
+        // post-conditions
+        assert(
+            naughtCoin.balanceOf(player) == playerBalanceBefore &&
+                naughtCoin.balanceOf(bob) == bobBalanceBefore
+        );
+    }
+
+   ```
+
+Run Echidna and check the corpus. Now we have fully covered this property.
+
+</details>
+
+<details>
+  <summary>4th property</summary>
+  
+  According to the ERC20 spec, the `approve` function should not fail if the caller has enough tokens to make the approval.
```

Comment:
`approve` won't fail even if you don't have enough tokens. It's like signing a blank check: "I allow this account to spend up to xxx of my funds in one or more transactions"
It might happen that I don't have the funds yet, but the approval goes through anyway.

Reply:
This is an oversight on my part. It's called an "allowance overrun", right? 

Will fix that. Thanks!

Reply:
Does it mean that this property is not the best one to test? If technically it will never fail?

Reply:
Fixed

Reply:
I'm not aware of it having a special name, the ERC20 standard does not mention it. You just allow the account to spend your funds.
Reviewing OZ's implementation, the only way it can fail is if the address zero is involved, otherwise it just updates the allowance.

Reply:
Fixed